Reporting db schema and build scripts for db and subscriber

eloaverona opened the pull request:
As part of the process of implementing the state delta export: 

- Adds schema for the reporting database.
- Adds docker image for postgres db and subscriber. 
- Establishes connection between subscriber and database.

Commit:
Add reporting database schema

Adds a representation of the reporting database that will be implemented
by the state delta export.

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

dplumb94 commented on subscriber/src/main.rs:
```diff
@@ -0,0 +1,66 @@
+        (@arg dbhost: default_value("localhost") --db_host +takes_value
+            "the host of the database")
+        (@arg dbport: default_value("5432") --db_port +takes_value
+            "the host of the database")
```
Small typo, should be "the port of the database"

eloaverona replied:
Fixed :)

dplumb94 commented:
I'd recommend splitting 057e712 up into 2 commits: one that implements the component boilerplate and one that takes care of the Docker deployment

dplumb94 commented on docker-compose.yaml:
```diff
@@ -30,3 +30,31 @@ services:
+    ports:
+      - '5432:5432'
+
+  sawtooth-bond-subscriber:
```
Recommend changing this to `bond-subscriber` to match other services

eloaverona replied:
Updated

Commit:
Add docker build scripts for subscriber and database

Creates the framework for the subscriber, including a Dockerfile.
Adds the subscriber and a postgress database to the main
docker-compose file.


Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Add boilerplate components for subscriber development

Adds a cli that establishes default dsn for db and takes
input about a postgres database to connect the subscriber to.

Adds a database object that handles stablishing connection
to the db.

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

eloaverona commented:
@dplumb94 I split the commit.

dplumb94 commented on subscriber/src/main.rs:
```diff
@@ -0,0 +1,64 @@
+
```
Extra line

dplumb94 commented on subscriber/src/main.rs:
```diff
@@ -0,0 +1,64 @@
+    database.connect().expect("Failed to connect to database");
+    database.disconnect().expect("Failed to close connection to database");
+
+
```
Extra lines

dplumb94 commented on subscriber/src/database.rs:
```diff
@@ -0,0 +1,50 @@
+    pub fn disconnect(&mut self) -> Result<(), Error>{
+        if self.conn.is_some() {
+            self.conn.take().unwrap().finish()?;
+
```
Extra line

dplumb94 commented on subscriber/Dockerfile:
```diff
@@ -0,0 +1,31 @@
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+# to build:
```
Is there a case where someone will need to build this by itself (not using docker-compose)? Could add confusion

RyanLassigBanks replied:
This could be useful if build dependencies were cached in the docker file or the subscriber itself was built in the image. That would improve `docker-compose up` times.

dplumb94 commented on subscriber/src/database.rs:
```diff
@@ -0,0 +1,50 @@
+        unimplemented!()
+    }
+
+    pub fn connect(&mut self) -> Result<(), Error> {
```
Does this do any sort of connection retry by default? If Postgres hasn't started yet, we could get a connection error

dplumb94 commented:
Got a few compilation warnings:
```
sawtooth-bond-subscriber    |    Compiling sawtooth_bond_subscriber v0.1.0 (file:///project/sawtooth-bond/subscriber)
sawtooth-bond-subscriber    | warning: unused imports: `Connection`, `TlsMode`
sawtooth-bond-subscriber    |   --> src/main.rs:23:16
sawtooth-bond-subscriber    |    |
sawtooth-bond-subscriber    | 23 | use postgres::{Connection, TlsMode};
sawtooth-bond-subscriber    |    |                ^^^^^^^^^^  ^^^^^^^
sawtooth-bond-subscriber    |    |
sawtooth-bond-subscriber    |    = note: #[warn(unused_imports)] on by default
sawtooth-bond-subscriber    |
sawtooth-bond-subscriber    | warning: unused variable: `logger`
sawtooth-bond-subscriber    |   --> src/main.rs:48:9
sawtooth-bond-subscriber    |    |
sawtooth-bond-subscriber    | 48 |     let logger = match matches.occurrences_of("verbose") {
sawtooth-bond-subscriber    |    |         ^^^^^^ help: consider using `_logger` instead
sawtooth-bond-subscriber    |    |
sawtooth-bond-subscriber    |    = note: #[warn(unused_variables)] on by default
sawtooth-bond-subscriber    |
sawtooth-bond-subscriber    | warning: method is never used: `create_tables`
sawtooth-bond-subscriber    |   --> src/database.rs:31:5
sawtooth-bond-subscriber    |    |
sawtooth-bond-subscriber    | 31 |     pub fn create_tables(&self) -> Result<(), Error> {
sawtooth-bond-subscriber    |    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sawtooth-bond-subscriber    |    |
sawtooth-bond-subscriber    |    = note: #[warn(dead_code)] on by default
sawtooth-bond-subscriber    |
sawtooth-bond-subscriber    |     Finished dev [unoptimized + debuginfo] target(s) in 3m 48.40s
```

I'd recommend removing (or using/testing) the unused code

RyanLassigBanks commented on subscriber/src/database.rs:
```diff
@@ -0,0 +1,50 @@
+extern crate postgres;
+use postgres::{Connection, TlsMode, Error};
+
+pub struct Database {
```
I'd recommend using [Diesel](https://github.com/diesel-rs/diesel) and [R2D2](https://github.com/sfackler/r2d2) to handle database connections and sending queries. This will give you things like database connection retries for free.

RyanLassigBanks commented on subscriber/src/database.rs:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    pub fn create_tables(&self) -> Result<(), Error>{
```
I don't think the subscriber should be responsible for creating tables. For rust implementations we usually have a `*.sql` script initialize the database when the postgres container starts up

eloaverona commented:
I closed this pull request and opened a [new one](https://github.com/bitwiseio/sawtooth-bond/pull/6), that is more specific to the the implementation of the reporting database.
